src/include/duckdb/common/union_by_name.hpp

```diff
@@ -49,28 +50,92 @@ class UnionByName {
 	static vector<typename READER_TYPE::UNION_READER_DATA>
 	UnionCols(ClientContext &context, const vector<string> &files, vector<LogicalType> &union_col_types,
 	          vector<string> &union_col_names, OPTION_TYPE &options) {
-		vector<typename READER_TYPE::UNION_READER_DATA> union_readers;
-		union_readers.resize(files.size());
-
-		TaskExecutor executor(context);
-		// schedule tasks for all files
-		for (idx_t file_idx = 0; file_idx < files.size(); ++file_idx) {
-			auto task = make_uniq<UnionByReaderTask<READER_TYPE, OPTION_TYPE>>(executor, context, files[file_idx],
-			                                                                   file_idx, union_readers, options);
-			executor.ScheduleTask(std::move(task));
-		}
-		// complete all tasks
-		executor.WorkOnTasks();
-
-		// now combine the result schemas
-		case_insensitive_map_t<idx_t> union_names_map;
-		for (auto &reader : union_readers) {
-			auto &col_names = reader->names;
-			auto &sql_types = reader->types;
-			CombineUnionTypes(col_names, sql_types, union_col_types, union_col_names, union_names_map);
+		if (!options.file_options.cache_union_readers) {
+			// Each worker thread sequentially unifies schemas for a subset of files
+			// without keeping readers in memory
+			idx_t max_threads = static_cast<idx_t>(TaskScheduler::GetScheduler(context).NumberOfThreads());
+			vector<vector<LogicalType>> union_col_types_per_thread(max_threads);
+			vector<vector<string>> union_col_names_per_thread(max_threads);
```

Comment:
Would it be possible to unify the code paths for when the union readers are cached and aren't cached?

If we save the types and names in these vectors in both cases:
```c++
vector<vector<LogicalType>> union_col_types_per_thread(max_threads);
vector<vector<string>> union_col_names_per_thread(max_threads);
```
Could we handle the logic for caching/not caching the reader within the task?

Then we only have one code path, and adding more options down the line becomes easier. Have you already tried this or think it's not possible?

Reply:
Thanks for the suggestion! I had not previously considered this and implementing it yielded much cleaner code.